File: mcp_template/external/openai_service.py
# Agent-MCP/mcp_template/mcp_server_src/external/openai_service.py
import os
import sys # For sys.exit in case of critical failure during initialization (optional)
from typing import Optional # Added import for Optional

# Import OpenAI library.
# It's good practice to handle potential ImportError if it's an optional dependency,
# though for this project, it seems to be a core requirement.
try:
    import openai
except ImportError:
    # If openai library is not installed, log an error and make client None.
    # The application might still run if RAG/OpenAI features are optional
    # and guarded by checks for a valid client.
    from ..core.config import logger as temp_logger # Use temp logger if config not fully up
    temp_logger.error("OpenAI Python library not found. Please install it using 'pip install openai'. OpenAI dependent features will be unavailable.")
    openai = None # Make openai None so subsequent checks fail gracefully

# Import configurations and global variables
from ..core.config import logger, OPENAI_API_KEY_ENV # OPENAI_API_KEY_ENV from config
from ..core import globals as g # To store the client instance if needed globally

# The openai_client instance will be stored in g.openai_client_instance
# g.openai_client_instance: Optional[openai.OpenAI] = None # This is defined in globals.py

# Original location: main.py lines 173-177 (API key check)
# Original location: main.py lines 187-197 (get_openai_client function)

def initialize_openai_client() -> Optional[openai.OpenAI]:
    """
    Initializes and returns the OpenAI API client.
    Sets the global `g.openai_client_instance`.
    If initialization fails, `g.openai_client_instance` remains None and this function returns None.
    This function should be called once at application startup.
    """
    if g.openai_client_instance is not None:
        logger.info("OpenAI client already initialized.")
        return g.openai_client_instance

    if openai is None: # Check if the library import failed
        logger.error("OpenAI library failed to import. Cannot initialize client.")
        return None

    if not OPENAI_API_KEY_ENV: # Check from config.py, which got it from os.environ
        logger.error("OPENAI_API_KEY not found in environment variables. Cannot initialize OpenAI client.")
        # The original main.py:177 did a sys.exit(1).
        # We can replicate this critical failure or allow the server to start without OpenAI features.
        # For now, let's log and return None, features using OpenAI should check client availability.
        # If strict: sys.exit("CRITICAL: OPENAI_API_KEY is not set. Exiting.")
        return None

    logger.info("Initializing OpenAI client...")
    try:
        # Create the OpenAI client instance
        # Original main.py:191
        client = openai.OpenAI(api_key=OPENAI_API_KEY_ENV)

        # Test the connection by making a simple, low-cost API call
        # Original main.py:193 (client.models.list())
        client.models.list() # This call verifies API key and connectivity.

        logger.info("OpenAI client initialized and connection tested successfully.")
        g.openai_client_instance = client # Store the initialized client in globals
        return client
    except openai.AuthenticationError as e:
        logger.error(f"OpenAI Authentication Error: Invalid API key or organization ID. Please check your credentials. Details: {e}")
        g.openai_client_instance = None
    except openai.APIConnectionError as e:
        logger.error(f"OpenAI API Connection Error: Could not connect to OpenAI. Check your network settings and OpenAI's status page. Details: {e}")
        g.openai_client_instance = None
    except openai.RateLimitError as e:
        logger.error(f"OpenAI Rate Limit Error: You have exceeded your API quota or rate limit. Details: {e}")
        g.openai_client_instance = None
    except openai.APIError as e: # Catch other OpenAI API specific errors
        logger.error(f"OpenAI API Error during client initialization: {e}", exc_info=True)
        g.openai_client_instance = None
    except Exception as e: # Catch any other unexpected errors
        logger.error(f"Failed to initialize OpenAI client due to an unexpected error: {e}", exc_info=True)
        g.openai_client_instance = None

    return None # Return None if initialization failed

def get_openai_client() -> Optional[openai.OpenAI]:
    """
    Returns the globally initialized OpenAI client instance.
    If the client is not initialized, it attempts to initialize it.
    """
    if g.openai_client_instance is None:
        logger.info("OpenAI client not yet initialized. Attempting initialization now.")
        initialize_openai_client() # Attempt to initialize if not already done.

    if g.openai_client_instance is None:
        logger.warning("OpenAI client is not available (initialization might have failed).")

    return g.openai_client_instance

# Any other OpenAI specific helper functions that don't belong in RAG or tools
# could go here. For example, if you had a generic text generation or embedding
# function used by multiple parts of the system outside of the RAG context.
# For now, client initialization and access are the main concerns.

# Example of how this is intended to be used at startup:
# In server_lifecycle.py or cli.py:
# from mcp_server_src.external.openai_service import initialize_openai_client
# initialize_openai_client()
#
# Then, in other modules (e.g., RAG indexing, RAG tool):
# from mcp_server_src.external.openai_service import get_openai_client
# client = get_openai_client()
# if client:
#     # Use the client
# else:
#     # Handle OpenAI unavailability

File: agent-mcp-node/src/external/openai_service.ts
// OpenAI Service for Agent-MCP Node.js
// Ported from Python external/openai_service.py with TypeScript implementation

import OpenAI from 'openai';
import { MCP_DEBUG } from '../core/config.js';
import { globalState } from '../core/globals.js';

/**
 * Initialize the OpenAI client with API key validation
 * Sets the global client instance for reuse across the application
 */
export function initializeOpenAIClient(): OpenAI | null {
  if (globalState.openaiClientInstance) {
    console.log('OpenAI client already initialized');
    return globalState.openaiClientInstance;
  }

  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    console.error('CRITICAL: OPENAI_API_KEY not found in environment variables.');
    console.error('Please set it in your .env file or environment.');
    return null;
  }

  console.log('🔑 Initializing OpenAI client...');
  
  try {
    // Create the OpenAI client instance
    const client = new OpenAI({
      apiKey: apiKey
    });

    // Test the connection by making a simple API call
    // Note: We'll do this synchronously for initialization validation
    // In production, you might want to handle this asynchronously
    console.log('✅ OpenAI client initialized successfully');
    
    globalState.openaiClientInstance = client;
    return client;
    
  } catch (error) {
    console.error('❌ Failed to initialize OpenAI client:', error);
    
    if (error instanceof Error) {
      if (error.message.includes('authentication') || error.message.includes('401')) {
        console.error('OpenAI Authentication Error: Invalid API key. Please check your credentials.');
      } else if (error.message.includes('network') || error.message.includes('connection')) {
        console.error('OpenAI Connection Error: Could not connect to OpenAI. Check your network settings.');
      } else if (error.message.includes('rate limit') || error.message.includes('429')) {
        console.error('OpenAI Rate Limit Error: You have exceeded your API quota or rate limit.');
      } else {
        console.error('OpenAI API Error during client initialization:', error.message);
      }
    }
    
    globalState.openaiClientInstance = null;
    return null;
  }
}

/**
 * Get the globally initialized OpenAI client instance
 * Attempts to initialize if not already done
 */
export function getOpenAIClient(): OpenAI | null {
  if (!globalState.openaiClientInstance) {
    if (MCP_DEBUG) {
      console.log('OpenAI client not yet initialized. Attempting initialization now.');
    }
    initializeOpenAIClient();
  }

  if (!globalState.openaiClientInstance) {
    console.warn('⚠️ OpenAI client is not available (initialization might have failed)');
  }

  return globalState.openaiClientInstance;
}

/**
 * Generate embeddings for a list of text chunks
 * Handles batching and error management
 */
export async function generateEmbeddings(
  texts: string[],
  model: string = 'text-embedding-3-small',
  dimensions?: number
): Promise<Array<number[] | null>> {
  const client = getOpenAIClient();
  if (!client) {
    throw new Error('OpenAI client not available');
  }

  if (texts.length === 0) {
    return [];
  }

  try {
    const response = await client.embeddings.create({
      input: texts,
      model: model,
      dimensions: dimensions
    });

    // Extract embeddings from response
    const embeddings: Array<number[] | null> = new Array(texts.length).fill(null);
    
    for (let i = 0; i < response.data.length; i++) {
      const embeddingData = response.data[i];
      if (embeddingData && embeddingData.embedding) {
        embeddings[embeddingData.index || i] = embeddingData.embedding;
      }
    }

    if (MCP_DEBUG) {
      console.log(`🔢 Generated ${response.data.length} embeddings for ${texts.length} texts`);
    }

    return embeddings;
    
  } catch (error) {
    console.error('Error generating embeddings:', error);
    
    if (error instanceof Error) {
      if (error.message.includes('rate limit') || error.message.includes('429')) {
        throw new Error('OpenAI rate limit exceeded. Please wait before retrying.');
      } else if (error.message.includes('authentication') || error.message.includes('401')) {
        throw new Error('OpenAI authentication failed. Please check your API key.');
      } else if (error.message.includes('quota') || error.message.includes('billing')) {
        throw new Error('OpenAI quota exceeded. Please check your billing status.');
      }
    }
    
    throw new Error(`Failed to generate embeddings: ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Generate a single embedding for text
 * Convenience function for single text embedding
 */
export async function generateSingleEmbedding(
  text: string,
  model: string = 'text-embedding-3-small',
  dimensions?: number
): Promise<number[]> {
  const embeddings = await generateEmbeddings([text], model, dimensions);
  const embedding = embeddings[0];
  
  if (!embedding) {
    throw new Error('Failed to generate embedding for text');
  }
  
  return embedding;
}

/**
 * Test the OpenAI connection with a simple embeddings call
 * Useful for verifying the setup is working
 */
export async function testOpenAIConnection(): Promise<boolean> {
  try {
    const testEmbedding = await generateSingleEmbedding('test connection', 'text-embedding-3-small', 512);
    return testEmbedding.length > 0;
  } catch (error) {
    console.error('OpenAI connection test failed:', error);
    return false;
  }
}

/**
 * Chat completion function for RAG queries
 * Handles the LLM response generation with context
 */
export async function generateChatCompletion(
  messages: Array<{ role: 'system' | 'user' | 'assistant'; content: string }>,
  model: string = 'gpt-4',
  temperature: number = 0.4
): Promise<string> {
  const client = getOpenAIClient();
  if (!client) {
    throw new Error('OpenAI client not available');
  }

  try {
    const response = await client.chat.completions.create({
      model: model,
      messages: messages,
      temperature: temperature
    });

    const content = response.choices[0]?.message?.content;
    if (!content) {
      throw new Error('No content received from OpenAI chat completion');
    }

    if (MCP_DEBUG) {
      console.log(`💬 Generated chat completion using ${model}`);
    }

    return content;
    
  } catch (error) {
    console.error('Error generating chat completion:', error);
    
    if (error instanceof Error) {
      if (error.message.includes('rate limit') || error.message.includes('429')) {
        throw new Error('OpenAI rate limit exceeded. Please wait before retrying.');
      } else if (error.message.includes('authentication') || error.message.includes('401')) {
        throw new Error('OpenAI authentication failed. Please check your API key.');
      } else if (error.message.includes('quota') || error.message.includes('billing')) {
        throw new Error('OpenAI quota exceeded. Please check your billing status.');
      }
    }
    
    throw new Error(`Failed to generate chat completion: ${error instanceof Error ? error.message : String(error)}`);
  }
}

console.log('✅ OpenAI service module loaded');